Mathlib/Tactic/Linter/DeprecatedModule.lean

```diff
@@ -0,0 +1,137 @@
+  if mm.getNat == 0 || 12 < mm.getNat || mm.raw.getSubstring?.get!.toString.trim.length != 2 then
+    throwErrorAt mm "The month should be of the form 01, 02, ..., 12!"
+  if dd.getNat == 0 || 31 < dd.getNat || dd.raw.getSubstring?.get!.toString.trim.length != 2 then
+    throwErrorAt dd "The day should be of the form 01, 02, ..., 31!"
+  addModuleDeprecation
+  -- disable the linter, so that it does not complain in the file with the deprecation
+  elabCommand (← `(set_option linter.deprecated.module false))
+
+/-- A utility function to show the pairings `(deprecatedModule, #[preferredModules])`. -/
+elab "#show_deprecated_modules" : command => do
+  let directImports := deprecatedModuleExt.getState (← getEnv)
+  logInfo <| "\n".intercalate <|
+    directImports.fold (init := ["Deprecated modules\n"]) fun nms (i, deps) =>
+      nms ++ [s!"'{i}' deprecates to\n{deps}\n"]
+
+namespace DeprecatedModule
+
+@[inherit_doc Mathlib.Linter.linter.deprecated.module]
+def deprecated.moduleLinter : Linter where run := withSetOptionIn fun stx ↦ do
+  let deprecations := deprecatedModuleExt.getState (← getEnv)
+  if deprecations.isEmpty then
+    return
+  unless Linter.getLinterValue linter.deprecated.module (← getOptions) do
+    return
+  if (← get).messages.hasErrors then
+    return
+  let mainModule ← getMainModule
+  -- `Mathlib.lean` and `Mathlib/Tactic.Lean` are allowed to import deprecated files,
+  -- since they (are expected to) import all modules in certain directories.
+  if #[`Mathlib, `Mathlib.Tactic].contains mainModule then return
+  if stx.isOfKind ``Linter.deprecated_modules then return
+  let fm ← getFileMap
+  let md := (getMainModuleDoc (← getEnv)).toArray
+  -- The end of the first module doc-string, or the end of the file if there is none.
+  let firstDocModPos := match md[0]? with
+                          | none     => fm.positions.back!
+                          | some doc => fm.ofPosition doc.declarationRange.endPos
+  unless stx.getTailPos?.getD default ≤ firstDocModPos do
+    return
+  -- We try to parse the file up to `firstDocModPos`.
+  let upToStx ← parseUpToHere firstDocModPos <|> (do
+    -- If parsing failed, there is some command which is not a module docstring.
+    -- In that case, we parse until the end of the modules and add an extra `section` afterwards,
+    -- so we trigger a "no module doc-string" warning.
+    let fil ← getFileName
+    let (stx, _) ← Parser.parseHeader { input := fm.source, fileName := fil, fileMap := fm }
+    parseUpToHere (stx.getTailPos?.getD default) "\nsection")
+  let importIds := getImportIds upToStx
+  let modulesWithNames := importIds.map fun i => (i, i.getId)
+  for is@(i, undeprecated) in deprecations do
+    for (nmStx, _) in modulesWithNames.filter (·.2 == i) do
+      Linter.logLint linter.deprecated.module nmStx
+        m!"'{nmStx}' has been deprecated: please replace this import by \
+          {"\nimport ".intercalate <| "\n" :: (undeprecated.map (·.toString)).toList}\n"
```

Comment:
```suggestion
          \n{"\n".intercalate <| (undeprecated.map (s"import {·.toString}")).toList}\n"
```
is a little less wonky

Reply:
I tried a `foldl` approach, avoiding intercalate.